subgraph/schema.graphql

```diff
@@ -56,3 +59,134 @@ type AuthorizedSigner @entity {
+}
+
+"""
+User created a net-new active Subscription
+"""
+type UserSubscriptionCreatedEvent implements UserSubscriptionsEvent
+  @entity(immutable: true) {
+  # keccak256 hex string of user:created:timestamp
+  id: Bytes!
+  user: User!
+  blockNumber: BigInt!
+  blockTimestamp: BigInt!
+  txHash: Bytes!
+  # will always be UserSubscriptionsEventType.CREATED
+  eventType: UserSubscriptionsEventType!
+  # the created active subscription
+  activeSubscription: ActiveSubscription!
+}
+
+"""
+User canceled their active subscription
+"""
+type UserSubscriptionCanceledEvent implements UserSubscriptionsEvent
+  @entity(immutable: true) {
+  # keccak256 hex string of user:canceled:timestamp
+  id: Bytes!
+  user: User!
+  blockNumber: BigInt!
+  blockTimestamp: BigInt!
+  txHash: Bytes!
+  # will always be UserSubscriptionsEventType.CANCELED
+  eventType: UserSubscriptionsEventType!
+  # the amount of unlocked tokens transferred back to the user as a result of cancelling the ActiveSubscription
+  tokensReturned: BigInt!
+}
+
+"""
+User renewed their active Subscription.
+
+Renewing means that the user extended the end timestamp of their active Subscription to a later timestamp.
+"""
+type UserSubscriptionRenewalEvent implements UserSubscriptionsEvent
+  @entity(immutable: true) {
+  # keccak256 hex string of user:renew:timestamp
+  id: Bytes!
+  user: User!
+  blockNumber: BigInt!
+  blockTimestamp: BigInt!
+  txHash: Bytes!
+  # will always be UserSubscriptionsEventType.RENEW
+  eventType: UserSubscriptionsEventType!
+  # the user's active subscription
+  activeSubscription: ActiveSubscription!
+}
+
+"""
+User upgraded their active Subscription.
+
+Upgrading means that the user increased the `rate` on their active Subscription, granting them more query volume.
+"""
+type UserSubscriptionUpgradeEvent implements UserSubscriptionsEvent
+  @entity(immutable: true) {
+  # keccak256 hex string of user:upgrade:timestamp
+  id: Bytes!
+  user: User!
+  blockNumber: BigInt!
+  blockTimestamp: BigInt!
+  txHash: Bytes!
+  # will always be UserSubscriptionsEventType.UPGRADE
+  eventType: UserSubscriptionsEventType!
+  # the user's active subscription
+  activeSubscription: ActiveSubscription!
+  # the previous ActiveSubscription.rate
+  previousRate: BigInt! # uint128
+}
+
+"""
+User downgraded their active Subscription.
+
+Downgrading means that the user decreased the `rate` on their active Subscription, granting them less query volume.
+"""
+type UserSubscriptionDowngradeEvent implements UserSubscriptionsEvent
+  @entity(immutable: true) {
+  # keccak256 hex string of user:downgrade:timestamp
+  id: Bytes!
+  user: User!
+  blockNumber: BigInt!
+  blockTimestamp: BigInt!
+  txHash: Bytes!
+  # will always be UserSubscriptionsEventType.DOWNGRADE
+  eventType: UserSubscriptionsEventType!
+  # the user's active subscription
+  activeSubscription: ActiveSubscription!
+  # the previous ActiveSubscription.rate
+  previousRate: BigInt! # uint128
```

Comment:
Won’t we need the start and end of the subscription at the time? The design a
expects a delta of full monthly rates I think (Or do we have it in activeSubscription)?

Reply:
I think the `activeSubscription` will be updated with a new `start` time, so we should probably store the prior `start`.

Reply:
I mentioned this above as well. But I think it makes sense to flatten the active subscription _before the update_ and then link to the current so that the deltas can be resolved for the event.